rig/routing_table/minimise.py

```diff
@@ -0,0 +1,557 @@
+"""Minimisation of SpiNNaker Routing Tables
+
+Background
+==========
+
+SpiNNaker routing tables consist of entries made up of a 32-bit key, a 32-bit
+mask and a 24-bit route value. The key and mask of every entry act as a sieve
+for the keys found on incoming multicast packets. Each bit of the key-mask pair
+can be considered as matching 0, 1 or 2 values in the same bit of a multicast
+packet key:
+
+ =====   =====   ==================   =======
+ Key     Mask    Matches Key Values   Written
+ =====   =====   ==================   =======
+ ``0``   ``0``   ``0`` or ``1``       ``X``
+ ``0``   ``1``   ``0``                ``0``
+ ``1``   ``1``   ``1``                ``1``
+ ``1``   ``0``   Nothing              ``!``
+ =====   =====   ==================   =======
+
+If a packet matches the key-mask of an entry then the packet is transmitted to
+the cores and links indicated by the route field.
+
+For example, if the table were:
+
+ =======   ========  =================
+ Key       Mask      Route
+ =======   ========  =================
+ ``0000``  ``1111``  North, North East
+ ``0111``  ``0111``  South
+ =======   ========  =================
+
+Which, from now on, will be written as::
+
+    0000 -> N NE
+    X111 -> S
+
+Then any packets with the key ``0000`` would be sent out of the north and
+north-east links. Any packets with the keys ``0111`` or ``1111`` would be sent
+out of the south link only.
+
+Entries in table are ordered, with entries at the top of the table having
+higher priority than those lower down the table. Only the highest priority
+entry which matches a packet is used. If, for example, the table were::
+
+    0000 -> N NE
+    1111 -> 1 2
+    X111 -> S
+
+Then packets with the keys ``0000`` and ``0111`` would be treated as before.
+However, packets with the key ``1111`` would be sent to cores 1 and 2 as only
+the higher priority entry has effect.
+
+Merging routing table entries
+=============================
+
+Routing tables can be minimised by merging together entries with equivalent
+routes. This is done by creating a new key-mask pair with an ``X`` wherever the
+key-mask pairs of any of the original entries differed.
+
+For example, merging of the entries::
+
+    0000 -> N
+    0001 -> N
+
+Would lead to the new entry:
+
+    000X -> N
+
+Which would match any of the keys matched by the original entries but no more.
+In contrast the merge of ``0001`` and ``0010`` would generate the new entry
+``00XX`` which would match keys matched by either of the original entries but
+also ``0000`` and ``0011``.
+
+Clearly, if we are to attempt to minimise tables such as::
+
+    0001 -> N
+    0010 -> N
+    0000 -> S, SE
+    0011 -> SE
+
+We need a set of rules for:
+
+ 1. Where merged entries are to be inserted into the table
+ 2. Which merges are allowed
+
+"Ordered Covering"
+==================
+
+The algorithm implemented here, "Ordered Covering", provides the following
+rules:
```

Comment:
I'm not sure I'd use the term "*generating*", but if the table you receive is in order of increasing generality and you apply these rules when you remove and add entries to the table then the final table, and every intermediate table, is guaranteed to be functionally equivalent to the starting table.

(*Note* if the starting table is ordered but is non-orthogonal then following these rules will still guarantee a functionally equivalent table but it may not be a small as it could be if you removed the non-orthogonal entries.)